ast/compile.go

```diff
@@ -1728,6 +1735,218 @@ func (c *Compiler) rewriteDynamicTerms() {
+				if isRegoMetadataChainCall(expr) || isRegoMetadataRuleCall(expr) {
+					regoMetadataCalled = true
+				}
+				return regoMetadataCalled
+			})
+			return regoMetadataCalled
+		})
+
+		if regoMetadataCalled {
+			break
+		}
+	}
+
+	if regoMetadataCalled {
+		// NOTE: Possible optimization: only parse annotations for modules on the path of rego.metadata-calling module
+		for _, name := range c.sorted {
+			mod := c.Modules[name]
+
+			if len(mod.Annotations) == 0 {
+				var errs Errors
+				mod.Annotations, errs = parseAnnotations(mod.Comments)
+				errs = append(errs, attachAnnotationsNodes(mod)...)
+				for _, err := range errs {
+					c.err(err)
+				}
+			}
+		}
+	}
+}
+
+func (c *Compiler) rewriteRegoMetadataCalls() {
+	for _, name := range c.sorted {
+		mod := c.Modules[name]
+
+		WalkRules(mod, func(rule *Rule) bool {
+			var regoMetadataChainCalled bool
+			var regoMetadataRuleCalled bool
+
+			WalkExprs(rule, func(expr *Expr) bool {
+				if isRegoMetadataChainCall(expr) {
+					regoMetadataChainCalled = true
+				} else if isRegoMetadataRuleCall(expr) {
+					regoMetadataRuleCalled = true
+				}
+				return regoMetadataChainCalled && regoMetadataRuleCalled
+			})
+
+			if regoMetadataChainCalled || regoMetadataRuleCalled {
+				body := make(Body, 0, len(rule.Body)+2)
+
+				var metadataChainVar Var
+				if regoMetadataChainCalled {
+					// Create and inject metadata chain for rule
+
+					chain, err := createMetadataChain(c.annotationSet.Chain(rule))
+					if err != nil {
+						c.err(err)
+						return false
+					}
+
+					f := newEqualityFactory(c.localvargen)
+					eq := f.Generate(chain)
+					metadataChainVar = eq.Operands()[0].Value.(Var)
+					body.Append(eq)
+				}
+
+				var metadataRuleVar Var
+				if regoMetadataRuleCalled {
+					// Create and inject metadata for rule
+
+					var metadataRuleTerm *Term
+
+					a := getPrimaryRuleAnnotations(c.annotationSet, rule)
+					if a != nil {
+						annotObj, err := a.toObject()
+						if err != nil {
+							c.err(err)
+							return false
+						}
+						metadataRuleTerm = NewTerm(*annotObj)
+					} else {
+						// If rule has no annotations, assign an empty object
+						metadataRuleTerm = ObjectTerm()
+					}
+
+					f := newEqualityFactory(c.localvargen)
+					eq := f.Generate(metadataRuleTerm)
+					metadataRuleVar = eq.Operands()[0].Value.(Var)
+					body.Append(eq)
+				}
+
+				for _, expr := range rule.Body {
+					body.Append(expr)
+				}
+				rule.Body = body
+
+				vis := func(b Body) bool {
+					for _, err := range rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, b, &c.RewrittenVars) {
+						c.err(err)
+					}
+					return false
+				}
+				WalkBodies(rule.Head, vis)
+				WalkBodies(rule.Body, vis)
+			}
+
+			return false
+		})
+	}
+}
+
+func getPrimaryRuleAnnotations(as *AnnotationSet, rule *Rule) *Annotations {
+	annots := as.GetRuleScope(rule)
+
+	if len(annots) == 0 {
+		return nil
+	}
+
+	if len(annots) > 1 {
```

Comment:
[nit] this check seems redundant. sort.SliceStable would surely do little harm on a one-element slice. WDYT?